### ISSUE TITLE
pubsub

### ISSUE THREAD

**Issue by wormen** [localForage/localForage#622]
prompt
in your library implemented pubsub?

if no, whether it is planned to add?
In the documentation is a description of what is possible to realize this is a links
https://developer.mozilla.org/en/docs/Web/Events/storage
https://developer.mozilla.org/ru/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API

http://tutorials.jenkov.com/html5/local-storage.html

**Comment by tofumatt** [comment]
There's an existing issue open for this, but we have not implemented it yet: #104.

**Comment by thgreasi** [comment]
If you need a pubsub like implementation then you can use https://github.com/localForage/localForage-observable . Keep in mind that you will not receive notifications of manipulations originating from different browser tabs/windows. Observables are like a pubsub broker + promises + generators + steroids. They are getting increasingly popular within the RxJS and angular 2 community. Waiting for your feedback.

**Comment by wormen** [comment]
made his own version pubsub, may also nice for you
in web workers also works

``` js
module.exports = class PubSub {

    constructor(channel) {
        this.isBrowser = (typeof window === `object`);

        this._store = {};
        this._store[channel] = {time: 0};

        this.channel = channel;
        this.listener = () => {};
    }

    get() {
        if(this.isBrowser)
            return JSON.parse(localStorage.getItem(this.channel)).msg;
        else
            return this._store[this.channel].msg;
    }

    set(data){

        let obj = {
            msg: data,
            time: Date.now()
        };

        if(this.isBrowser)
            localStorage.setItem(this.channel, JSON.stringify(obj));
        else
            this._store[this.channel] = obj;
    }

    publish(data) {
        this.set(data);
        this.listener(this.get());
    }

    subscribe(listener) {
        this.listener = listener;

        if(this.isBrowser)
            window.addEventListener('storage', e => {
                if (e.key === this.channel) this.listener(this.get());
            }, false);
    }
};
```

``` js
let Channel = new PubSub('my_channel');

// Easily subscribe, data provided via callback.
Channel.subscribe(data => {
    console.log(data);
});

// Publish a new message to channel.
Channel.publish({message: 'Hello world!'});
```

**Comment by thgreasi** [comment]
Do you consider this as resolved? Should we close this issue?

**Comment by thgreasi** [comment]
Oh, didn't see that it's already closed.
Anyway, with the localforage.observable you have the advantage that you can continue using the existing localforage API instead of refactoring your code to use a wrapper.
I'm glad that you hacked a simple solution around localforage that worked fo your use case.
Kudos and thanks for sharing.

**Comment by wormen** [comment]
@thgreasi you meant https://github.com/localForage/localForage-observable ?
if so, how I can be used in conjunction with webpack, there is no information about this in the documentation

**Comment by thgreasi** [comment]
Just `require()` it or `import` it (depending on the way you import your libs) and it should extend the localforage prototype and make the `newObservable()` method available.
If you end up having teh `newObservable()` method `undefined` then try:
```js
import { extendPrototype } from 'localforage-observable';
extendPrototype(localforage);
```
or with `require()`:
```js
var lfObservable = require('localforage-observable');
lfObservable .extendPrototype(localforage);
```
Please comment back with more info on your setup if you face any issue.

**Comment by wormen** [comment]
@thgreasi 
when connected
``` js
let localforage = require('localforage');
let lfObservable  = require(`localforage-observable`);
lfObservable.extendPrototype(localforage);
```

I receive an error
`
Uncaught ReferenceError: localforage is not defined
`

**Comment by wormen** [comment]
1) in file localForage-observable/dist/localforage-observable.es6.js
changed
```
var extendPrototypeResult = extendPrototype(localforage);
```
to

```
var extendPrototypeResult = null;
if(typeof localforage != `undefined`)
    extendPrototypeResult = extendPrototype(localforage);
```

2) Total connection, without the appearance of errors, looks so

``` js
const localforage = require('localforage');

import Rx from 'rxjs/Rx';
const lfObservable  = require(`localforage-observable/dist/localforage-observable.es6`);

lfObservable.extendPrototype(localforage);

localforage.newObservable.factory = function (subscribeFn) {
    return Rx.Observable.create(subscribeFn);
};
```

3) the desired result has not received
data storage monitoring did not appear in the console

**Comment by wormen** [comment]
I published a file that you can see how I was doing
https://github.com/wormen/offline-storage/blob/master/store.js

**Comment by thgreasi** [comment]
Thanks! I will try to check it out during the weekend and possibly cut a new release as well.

**Comment by thgreasi** [comment]
Just released v1.0.0 of localforage-observable.
Please check whether it resolved the issues that you mentioned.
Thanks for your feedback so far!

**Comment by wormen** [comment]
@thgreasi 
after packing webpack
so nothing is displayed

**Comment by thgreasi** [comment]
I didn't get what you meant :(
Can you rephrase?
Does v1.0.0 of localforage-observable works as-is without errors for you?

**Comment by wormen** [comment]
@thgreasi
errors in the last release is not observed
but the problem has remained from the previous release

that the data or not tracked, according to documentation or does not fulfill a conclusion that the data has changed

I understandably described to you the essence of the problem?

**Comment by thgreasi** [comment]
> errors in the last release is not observed

Great!

> but the problem has remained from the previous release

That's logical...

> the data or not tracked,

That's strange. Can you fork [one of the provided exampes](http://codepen.io/thgreasi/pen/wGLWmv) to demonstrate your issue?
I will try to create a complete webpack specific test repo during the weekend.

**Comment by wormen** [comment]
@thgreasi
I previously gave a link to a file that pack in webpack
You can check on this file

https://github.com/wormen/offline-storage/blob/master/store.js

**Comment by thgreasi** [comment]
Can you chech what happend when using `const lfObservable  = require('localforage-observable');` and removing the `lfObservable.extendPrototype(localforage);` on v1.0.0 ?

**Comment by wormen** [comment]
error stopped appearing in the console
everything else remains unchanged

**Comment by thgreasi** [comment]
[Here](http://codepen.io/anon/pen/XNjzvX) you go. Had to fix some issues with the key prefixing.
On other thing to keep in mind is that the obsrervable notifies the subscribers only when the value actually changes. Otherwise, you can [disable the value checking](https://github.com/localForage/localForage-observable#value-change-detection).

**Comment by wormen** [comment]
@thgreasi thank you very much, now it is clear that I am not the right thing did

**Comment by thgreasi** [comment]
Just opened localForage/localForage-observable#6 for this. Any feedback would be great and could help releasing it during the weekend.

**Comment by gormonn** [comment]
hey

Why not just use localstorage to store hashes of stored objects in indexedDB?

After writing data to indexedDb - write key and hege values to localStorage
Accordingly, the other window will not receive an update if the hash has not changed.
If the hash has changed, we request data from indexedDb and do things
